File: support-utils/src/main/java/com/hendraanggrian/support/utils/content/Uris.java
/*
 * Copyright (c) 2015-present, Facebook, Inc.
 * All rights reserved.
 *
 * This source code is licensed under the BSD-style license found in the
 * LICENSE file in the root directory of this source tree. An additional grant
 * of patent rights can be found in the PATENTS file in the same directory.
 */

package com.hendraanggrian.support.utils.content;

import android.content.ContentResolver;
import android.content.Context;
import android.database.Cursor;
import android.net.Uri;
import android.provider.ContactsContract;
import android.provider.MediaStore;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;

/**
 * Taken from https://github.com/facebook/fresco/blob/master/fbcore/src/main/java/com/facebook/common/util/UriUtil.java.
 *
 * @author Hendra Anggrian (hendraanggrian@gmail.com)
 */
public final class Uris {

    public static final String SCHEME_HTTP = "http";
    public static final String SCHEME_HTTPS = "https";
    public static final String SCHEME_LOCAL_FILE = "file";
    public static final String SCHEME_LOCAL_CONTENT = "content";
    public static final String SCHEME_LOCAL_ASSET = "asset";
    public static final String SCHEME_LOCAL_RESOURCE = "res";
    public static final String SCHEME_QUALIFIED_RESOURCE = ContentResolver.SCHEME_ANDROID_RESOURCE;
    public static final String SCHEME_DATA = "data";

    private Uris() {
    }

    public static boolean isNetworkUri(@NonNull Uri uri) {
        String scheme = uri.getScheme();
        return SCHEME_HTTPS.equals(scheme) || SCHEME_HTTP.equals(scheme);
    }

    public static boolean isLocalFileUri(@NonNull Uri uri) {
        String scheme = uri.getScheme();
        return SCHEME_LOCAL_FILE.equals(scheme);
    }

    public static boolean isLocalContentUri(@NonNull Uri uri) {
        String scheme = uri.getScheme();
        return SCHEME_LOCAL_CONTENT.equals(scheme);
    }

    public static boolean isLocalContactUri(Uri uri) {
        return isLocalContentUri(uri)
                && ContactsContract.AUTHORITY.equals(uri.getAuthority())
                && !uri.getPath().startsWith(Uri.withAppendedPath(ContactsContract.AUTHORITY_URI, "display_photo").getPath());
    }

    public static boolean isLocalCameraUri(Uri uri) {
        String uriString = uri.toString();
        return uriString.startsWith(MediaStore.Images.Media.EXTERNAL_CONTENT_URI.toString())
                || uriString.startsWith(MediaStore.Images.Media.INTERNAL_CONTENT_URI.toString());
    }

    public static boolean isLocalAssetUri(@NonNull Uri uri) {
        String scheme = uri.getScheme();
        return SCHEME_LOCAL_ASSET.equals(scheme);
    }

    public static boolean isLocalResourceUri(@NonNull Uri uri) {
        String scheme = uri.getScheme();
        return SCHEME_LOCAL_RESOURCE.equals(scheme);
    }

    public static boolean isQualifiedResourceUri(@NonNull Uri uri) {
        String scheme = uri.getScheme();
        return SCHEME_QUALIFIED_RESOURCE.equals(scheme);
    }

    public static boolean isDataUri(@NonNull Uri uri) {
        return SCHEME_DATA.equals(uri.getScheme());
    }

    @Nullable
    public static String getActualPath(@NonNull Context context, Uri uri) {
        return getActualPath(context.getContentResolver(), uri);
    }

    @Nullable
    public static String getActualPath(@NonNull ContentResolver contentResolver, Uri uri) {
        String result = null;
        if (isLocalContentUri(uri)) {
            Cursor cursor = null;
            try {
                cursor = contentResolver.query(uri, null, null, null, null);
                if (cursor != null && cursor.moveToFirst()) {
                    int idx = cursor.getColumnIndex(MediaStore.Images.ImageColumns.DATA);
                    if (idx != -1)
                        result = cursor.getString(idx);
                }
            } finally {
                if (cursor != null)
                    cursor.close();
            }
        } else if (isLocalFileUri(uri)) {
            result = uri.getPath();
        }
        return result;
    }

    @NonNull
    public static Uri fromResourceId(int resourceId) {
        return new Uri.Builder()
                .scheme(SCHEME_LOCAL_RESOURCE)
                .path(String.valueOf(resourceId))
                .build();
    }

    @NonNull
    public static Uri fromQualifiedResource(@NonNull Context context, int resourceId) {
        return fromQualifiedResource(context.getPackageName(), resourceId);
    }

    @NonNull
    public static Uri fromQualifiedResource(@NonNull String packageName, int resourceId) {
        return new Uri.Builder()
                .scheme(SCHEME_QUALIFIED_RESOURCE)
                .authority(packageName)
                .path(String.valueOf(resourceId))
                .build();
    }
}

File: src/com/hendraanggrian/kota/net/Uris.kt
/*
 * Copyright (c) 2015-present, Facebook, Inc.
 * All rights reserved.
 *
 * This source code is licensed under the BSD-style license found in the
 * LICENSE file in the root directory of this source tree. An additional grant
 * of patent rights can be found in the PATENTS file in the same directory.
 */

package com.hendraanggrian.kota.net

import android.content.ContentResolver
import android.content.Context
import android.database.Cursor
import android.net.Uri
import android.provider.ContactsContract
import android.provider.MediaStore

object Uris {

    const val SCHEME_HTTP = "http"
    const val SCHEME_HTTPS = "https"
    const val SCHEME_FILE = ContentResolver.SCHEME_FILE
    const val SCHEME_CONTENT = ContentResolver.SCHEME_CONTENT
    const val SCHEME_ASSET = "asset"
    const val SCHEME_RESOURCE = "res"
    const val SCHEME_QUALIFIED_RESOURCE = ContentResolver.SCHEME_ANDROID_RESOURCE
    const val SCHEME_DATA = "data"
}

fun Uri.isNetwork(): Boolean = scheme.let { it == Uris.SCHEME_HTTPS || it == Uris.SCHEME_HTTP }

fun Uri.isFile(): Boolean = scheme == Uris.SCHEME_FILE

fun Uri.isContent(): Boolean = scheme == Uris.SCHEME_CONTENT

fun Uri.isContact(): Boolean = isContent()
        && ContactsContract.AUTHORITY == authority
        && !path.startsWith(Uri.withAppendedPath(ContactsContract.AUTHORITY_URI, "display_photo").path)

fun Uri.isCamera(): Boolean = toString().let { it.startsWith(MediaStore.Images.Media.EXTERNAL_CONTENT_URI.toString()) || it.startsWith(MediaStore.Images.Media.INTERNAL_CONTENT_URI.toString()) }

fun Uri.isAsset(): Boolean = scheme == Uris.SCHEME_ASSET

fun Uri.isResource(): Boolean = scheme == Uris.SCHEME_RESOURCE

fun Uri.isQualifiedResource(): Boolean = scheme == Uris.SCHEME_QUALIFIED_RESOURCE

fun Uri.isData(): Boolean = scheme == Uris.SCHEME_DATA

fun Uri.getActualPath(context: Context): String? = getActualPath(context.contentResolver)
fun Uri.getActualPath(resolver: ContentResolver): String? {
    var result: String? = null
    if (isContent()) {
        var cursor: Cursor? = null
        try {
            cursor = resolver.query(this, null, null, null, null)
            if (cursor != null && cursor.moveToFirst()) {
                val idx = cursor.getColumnIndex(MediaStore.Images.ImageColumns.DATA)
                if (idx != -1)
                    result = cursor.getString(idx)
            }
        } finally {
            if (cursor != null)
                cursor.close()
        }
    } else if (isFile()) {
        result = path
    }
    return result
}

/**
 * Build Uri from resource id.
 */
fun Int.toUri(): Uri = Uri.Builder()
        .scheme(Uris.SCHEME_RESOURCE)
        .path(toString())
        .build()

/**
 * Build Uri from qualified resource id.
 */
fun Int.toUri(context: Context): Uri = toUri(context.packageName)

fun Int.toUri(packageName: String): Uri = Uri.Builder()
        .scheme(Uris.SCHEME_QUALIFIED_RESOURCE)
        .authority(packageName)
        .path(toString())
        .build()